src/coprocessor/Client.cc

```diff
@@ -35,24 +44,377 @@ std::vector<copTask> buildCopTasks(
+            store_task_map[store_id].region_infos.emplace_back(region_info);
+            total_remaining_region_num -= 1;
+            for (const auto other_store_id : region_info.all_stores)
+            {
+                if (auto iter = store_candidate_region_map.find(other_store_id); iter != store_candidate_region_map.end())
+                {
+                    iter->second.erase(task_key);
+                    total_region_candidate_num -= 1;
+                    if (iter->second.empty())
+                        store_candidate_region_map.erase(iter);
+                }
+            }
+            if (total_remaining_region_num > 0)
+            {
+                avg_store_per_region = 1.0 * total_region_candidate_num / total_remaining_region_num;
+                store_id = find_next_store(region_info.all_stores);
+            }
+        }
+        if (total_remaining_region_num > 0)
+        {
+            log->warning("Some regions are not used when trying to balance batch cop task, give up balancing");
+            return std::move(original_tasks);
+        }
+    }
+
+    std::vector<BatchCopTask> ret;
+    ret.reserve(store_task_map.size());
+    for (const auto & [store_id, task] : store_task_map)
+    {
+        (void)store_id;
+        if (!task.region_infos.empty())
+        {
+            ret.emplace_back(task);
+        }
+    }
+    return ret;
+}
+} // namespace details
+
+// The elements in the two `physical_table_ids` and `ranges_for_each_physical_table` should be in one-to-one mapping.
+// When build batch cop tasks for partition table, physical_table_ids.size() may be greater than 1.
+std::vector<BatchCopTask> buildBatchCopTasks(
+    kv::Backoffer & bo,
+    kv::Cluster * cluster,
+    bool is_partition_table_scan,
+    const std::vector<int64_t> & physical_table_ids,
+    const std::vector<KeyRanges> & ranges_for_each_physical_table,
+    kv::StoreType store_type,
+    Logger * log)
+{
+    auto & cache = cluster->region_cache;
+    assert(physical_table_ids.size() == ranges_for_each_physical_table.size());
+
+    while (true) // for `need_retry`
+    {
+        std::vector<CopTask> cop_tasks;
+        for (size_t idx = 0; idx < ranges_for_each_physical_table.size(); ++idx)
+        {
+            const auto & ranges = ranges_for_each_physical_table[idx];
+            const auto locations = details::splitKeyRangesByLocations(cache, bo, ranges);
+            for (const auto & loc : locations)
+            {
+                cop_tasks.emplace_back(CopTask{loc.location.region, loc.ranges, nullptr, store_type, idx, kv::GRPCMetaData{}});
+            }
+        }
+
+        // store_addr -> BatchCopTask
+        std::map<std::string, BatchCopTask> store_task_map;
+        bool need_retry = false;
+        for (const auto & cop_task : cop_tasks)
+        {
+            auto rpc_context = cluster->region_cache->getRPCContext(bo, cop_task.region_id, store_type, false);
```

Comment:
Add some comments why `load_balance` is false here

Reply:
Done